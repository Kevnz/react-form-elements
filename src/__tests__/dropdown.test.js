import React from 'react'
import { render, fireEvent, cleanup } from '@testing-library/react'

import { DropDown, Option, OptionGroup } from '../'

describe('DropDown component', () => {
  afterEach(() => {
    cleanup()
  })
  describe('The Rendered HTML', () => {
    it('should render the full tree', () => {
      const { container } = render(
        <DropDown label="Drop Down" initialValue="2">
          <OptionGroup label="First Group">
            <Option initialValue="1">First</Option>
            <Option initialValue="2">Second</Option>
            <Option initialValue="3" label="Third" />
          </OptionGroup>
          <OptionGroup label="Second Group">
            <Option initialValue="11">Second First</Option>
            <Option initialValue="12">Second Second</Option>
            <Option initialValue="13" label="Second Third" />
          </OptionGroup>
          <OptionGroup label="Third Group">
            <Option initialValue="21">Third First</Option>
            <Option initialValue="22">Third Second</Option>
            <Option initialValue="23" label="Third Third" />
          </OptionGroup>
        </DropDown>
      )

      expect(container.firstChild).toMatchSnapshot()
    })

    it('should not need the OptionGroup', () => {
      const { container } = render(
        <DropDown label="Drop Down" initialValue="13">
          <Option initialValue="1">First</Option>
          <Option initialValue="2">Second</Option>
          <Option initialValue="3" label="Third" />
          <Option initialValue="11">Second First</Option>
          <Option initialValue="12">Second Second</Option>
          <Option initialValue="13" label="Second Third" />
          <Option initialValue="21">Third First</Option>
          <Option initialValue="22">Third Second</Option>
          <Option initialValue="23" label="Third Third" />
        </DropDown>
      )
      expect(container.firstChild).toMatchSnapshot()
    })
    it('should allow a mix of OptionGroups and Options', () => {
      const { container } = render(
        <DropDown label="Drop Down" initialValue="2" data-testid="dd1">
          <OptionGroup label="First Group">
            <Option initialValue="1">First</Option>
            <Option initialValue="2">Second</Option>
            <Option initialValue="3" label="Third" />
          </OptionGroup>
          <OptionGroup label="Second Group">
            <Option initialValue="11">Second First</Option>
            <Option initialValue="12">Second Second</Option>
            <Option initialValue="13" label="Second Third" />
          </OptionGroup>
          <Option initialValue="21">Third First</Option>
          <Option initialValue="22">Third Second</Option>
          <Option initialValue="23" label="Third Third" />
        </DropDown>
      )
      expect(container.firstChild).toMatchSnapshot()
    })
  })
  describe('handling change', () => {
    it('should set a value when passed in', () => {
      const { getByTestId } = render(
        <DropDown label="Drop Down" initialValue="2" data-testid="dd1">
          <OptionGroup label="First Group">
            <Option initialValue="1">First</Option>
            <Option initialValue="2">Second</Option>
            <Option initialValue="3" label="Third" />
          </OptionGroup>
          <OptionGroup label="Second Group">
            <Option initialValue="11">Second First</Option>
            <Option initialValue="12">Second Second</Option>
            <Option initialValue="13" label="Second Third" />
          </OptionGroup>
          <OptionGroup label="Third Group">
            <Option initialValue="21">Third First</Option>
            <Option initialValue="22">Third Second</Option>
            <Option initialValue="23" label="Third Third" />
          </OptionGroup>
        </DropDown>
      )
      const select = getByTestId('dd1')
      expect(select.value).toBe('2')
      fireEvent.change(select, { target: { value: '13' } })
      expect(select.value).toBe('13')
    })

    it('should set a value when passed in', () => {
      const { getByTestId } = render(
        <DropDown label="Drop Down" initialValue="13" data-testid="dd1">
          <Option initialValue="1">First</Option>
          <Option initialValue="2">Second</Option>
          <Option initialValue="3" label="Third" />
          <Option initialValue="11">Second First</Option>
          <Option initialValue="12">Second Second</Option>
          <Option initialValue="13" label="Second Third" />
          <Option initialValue="21">Third First</Option>
          <Option initialValue="22">Third Second</Option>
          <Option initialValue="23" label="Third Third" />
        </DropDown>
      )
      const select = getByTestId('dd1')
      expect(select.value).toBe('13')
      fireEvent.change(select, { target: { value: '1' } })
      expect(select.value).toBe('1')
    })
  })
})
