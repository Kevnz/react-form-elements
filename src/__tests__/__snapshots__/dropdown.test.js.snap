// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown component The Rendered HTML should allow a mix of OptionGroups and Options 1`] = `
<div
  class="rfe-form__row"
>
  <label
    for="id-3-1"
  >
    Drop Down
  </label>
  <select
    data-testid="dd1"
    id="id-3-1"
  >
    <optgroup
      label="First Group"
    >
      <option
        value="1"
      >
        First
      </option>
      <option
        value="2"
      >
        Second
      </option>
      <option
        value="3"
      >
        Third
      </option>
    </optgroup>
    <optgroup
      label="Second Group"
    >
      <option
        value="11"
      >
        Second First
      </option>
      <option
        value="12"
      >
        Second Second
      </option>
      <option
        value="13"
      >
        Second Third
      </option>
    </optgroup>
    <option
      value="21"
    >
      Third First
    </option>
    <option
      value="22"
    >
      Third Second
    </option>
    <option
      value="23"
    >
      Third Third
    </option>
  </select>
</div>
`;

exports[`DropDown component The Rendered HTML should not need the OptionGroup 1`] = `
<div
  class="rfe-form__row"
>
  <label
    for="id-2-1"
  >
    Drop Down
  </label>
  <select
    id="id-2-1"
  >
    <option
      value="1"
    >
      First
    </option>
    <option
      value="2"
    >
      Second
    </option>
    <option
      value="3"
    >
      Third
    </option>
    <option
      value="11"
    >
      Second First
    </option>
    <option
      value="12"
    >
      Second Second
    </option>
    <option
      value="13"
    >
      Second Third
    </option>
    <option
      value="21"
    >
      Third First
    </option>
    <option
      value="22"
    >
      Third Second
    </option>
    <option
      value="23"
    >
      Third Third
    </option>
  </select>
</div>
`;

exports[`DropDown component The Rendered HTML should render the full tree 1`] = `
<div
  class="rfe-form__row"
>
  <label
    for="id-1-1"
  >
    Drop Down
  </label>
  <select
    id="id-1-1"
  >
    <optgroup
      label="First Group"
    >
      <option
        value="1"
      >
        First
      </option>
      <option
        value="2"
      >
        Second
      </option>
      <option
        value="3"
      >
        Third
      </option>
    </optgroup>
    <optgroup
      label="Second Group"
    >
      <option
        value="11"
      >
        Second First
      </option>
      <option
        value="12"
      >
        Second Second
      </option>
      <option
        value="13"
      >
        Second Third
      </option>
    </optgroup>
    <optgroup
      label="Third Group"
    >
      <option
        value="21"
      >
        Third First
      </option>
      <option
        value="22"
      >
        Third Second
      </option>
      <option
        value="23"
      >
        Third Third
      </option>
    </optgroup>
  </select>
</div>
`;
